units/bson_to_parquet_backfill.go

```diff
@@ -0,0 +1,183 @@
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// TODO (EVG-16138): Remove this job once we do the BSON to Parquet cutover.
+
+const bsonToParquetBackfillJobName = "bson-to-parquet-backfill"
+
+type bsonToParquetBackfillJob struct {
+	job.Base `bson:"metadata" json:"metadata" yaml:"metadata"`
+
+	env         cedar.Environment
+	queue       amboy.Queue
+	controller  *model.BatchJobController
+	migrationID string
+}
+
+func init() {
+	registry.AddJobType(bsonToParquetBackfillJobName, func() amboy.Job { return makeBSONToParquetBackfillJob() })
+}
+
+func makeBSONToParquetBackfillJob() *bsonToParquetBackfillJob {
+	return &bsonToParquetBackfillJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    bsonToParquetBackfillJobName,
+				Version: 1,
+			},
+		},
+	}
+}
+
+// NewBSONToParquetBackfillJob returns a new Amboy job that does batched
+// migrations of existing BSON test results to Apache Parquet test results
+// stored in the Presto bucket.
+func NewBSONToParquetBackfillJob(id string) amboy.Job {
+	j := makeBSONToParquetBackfillJob()
+	j.SetID(fmt.Sprintf("%s.%s", bsonToParquetBackfillJobName, id))
+	j.SetEnqueueAllScopes(true)
+
+	return j
+}
+
+func (j *bsonToParquetBackfillJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+
+	if j.env == nil {
+		j.env = cedar.GetEnvironment()
+	}
+	if j.queue == nil {
+		j.queue = j.env.GetRemoteQueue()
+	}
+
+	controller, err := model.FindBatchJobController(ctx, j.env, bsonToParquetBackfillJobName)
+	if err != nil {
+		j.AddError(errors.Wrap(err, "getting batch job controller"))
+		return
+	}
+	if controller.Iterations <= 0 {
+		controller.Iterations = 1
+	}
+	j.controller = controller
+
+	if j.controller.Timeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, j.controller.Timeout)
+		defer cancel()
+	}
+
+	for i := 0; i < j.controller.Iterations; i++ {
+		if err = j.runIteration(ctx); err != nil {
+			j.AddError(err)
+			return
+		}
+	}
+}
+
+func (j *bsonToParquetBackfillJob) runIteration(ctx context.Context) error {
+	start := time.Now().UTC()
+	collection := j.env.GetDB().Collection(j.controller.Collection)
+
+	// This is really ugly, but unfortunately Mongo does not allow setting
+	// limits on UpdateMany operations so we have to do a BulkWrite with
+	// the batch size number of UpdateOne operations.
+	updates := make([]mongo.WriteModel, j.controller.BatchSize)
+	for i := 0; i < j.controller.BatchSize; i++ {
+		updates[i] = mongo.NewUpdateOneModel().
+			SetFilter(bson.M{"$or": []bson.M{
+				{model.TestResultsMigrationKey: bson.M{"$exists": false}},
+				{bsonutil.GetDottedKeyName(model.TestResultsMigrationKey, model.MigrationStatsVersionKey): bson.M{"$lt": j.controller.Version}},
+				{"$and": []bson.M{
+					{bsonutil.GetDottedKeyName(model.TestResultsMigrationKey, model.MigrationStatsStartedAtKey): bson.M{"$and": []bson.M{
+						{"gt": time.Time{}},
```

Comment:
this gets rows with started at > now?
and now is evaluated for each item?

Reply:
That is zero time, so basically this is looking for orphaned test results (test results that were marked for migration but the job failed before they could be marked complete).